File: src/main/java/leetcode/leetcode75/LC121.java
package leetcode.leetcode75;

/**
 * 121. Best Time to Buy and Sell Stock
 */
public class LC121 {

    /**
     * You are given an array prices where prices[i] is the price of a given stock on the ith day.
     * <p>
     * You want to maximize your profit by choosing a single day to buy one stock and choosing a different day in the future to sell that stock.
     * <p>
     * Return the maximum profit you can achieve from this transaction. If you cannot achieve any profit, return 0.
     * <p>
     * <p>
     * <p>
     * Example 1:
     * <p>
     * Input: prices = [7,1,5,3,6,4]
     * Output: 5
     * Explanation: Buy on day 2 (price = 1) and sell on day 5 (price = 6), profit = 6-1 = 5.
     * Note that buying on day 2 and selling on day 1 is not allowed because you must buy before you sell.
     * Example 2:
     * <p>
     * Input: prices = [7,6,4,3,1]
     * Output: 0
     * Explanation: In this case, no transactions are done and the max profit = 0.
     * <p>
     * <p>
     * Constraints:
     * <p>
     * 1 <= prices.length <= 105
     * 0 <= prices[i] <= 104
     *
     * @param args
     */
    public static void main(String[] args) {
        System.out.println(maxProfit(new int[]{2,4,1}));
    }

    public static int maxProfit(int[] prices) {
        int min = Integer.MAX_VALUE, max = 0;
        for (int i = 0; i < prices.length; i++) {
            if (prices[i] < min) {
                min = prices[i];
            } else if (prices[i] - min > max) {
                max = prices[i] - min;
            }
        }
        return max;
    }

}

File: src/main/java/leetcode/datastructure/day/three/LC121.kt
package leetcode.datastructure.day.three

/**
 * 121. Best Time to Buy and Sell Stock
 */
fun main(args: Array<String>) {
    println(maxProfit(intArrayOf(2, 1, 2, 1, 0, 1, 2)))
}

/**
 * 가장큰 값만구하면 되기 떄문에 최대 이득을 볼수있는 값을 미리 계산후 다음 min에서 들어오는 놈들의 이익 값을 비교해서 max값을 바꿈
 * 짜증난다
 */
fun maxProfit(prices: IntArray): Int {
    var min = Int.MAX_VALUE
    var max = 0
    for (i in 0 until prices.size) {
        if (prices[i] < min) {
            min = prices[i]
        } else if (prices[i] - min > max) {
            max = prices[i] - min
        }
    }
    return max
}